docker/configure_workers_and_start.py

```diff
@@ -362,19 +380,137 @@ def add_worker_roles_to_shared_config(
+            # work, this is fine.
+            new_dict[i] = existing_dict[i] + to_be_merged_dict[i]
+        else:
+            # Everything else should be identical, like "app", which only works because
+            # all apps are now generic_workers.
+            new_dict[i] = to_be_merged_dict[i]
+    return new_dict
+
+
+def insert_worker_name_for_worker_config(
+    existing_dict: Dict[str, Any], worker_name: str
+) -> Dict[str, Any]:
+    """Insert a given worker name into the worker's configuration dict.
+
+    Args:
+        existing_dict: The worker_config dict that is imported into shared_config.
+        worker_name: The name of the worker to insert.
+    Returns: Copy of the dict with newly inserted worker name
+    """
+    dict_to_edit = existing_dict.copy()
+    for k, v in dict_to_edit["shared_extra_conf"].items():
+        # Only proceed if it's a string as some values are boolean
+        if isinstance(dict_to_edit["shared_extra_conf"][k], str):
+            # This will be ignored if the text isn't the placeholder.
+            dict_to_edit["shared_extra_conf"][k] = v.replace(
+                "placeholder_name", worker_name
+            )
+    return dict_to_edit
+
+
+def is_sharding_allowed_for_worker_type(worker_type: str) -> bool:
+    """Helper to check to make sure worker types that cannot have multiples do not.
+
+    Args:
+        worker_type: The type of worker to check against.
+    Returns: True if allowed, False if not
+    """
+    if worker_type in [
+        "background_worker",
+        "account_data",
+        "presence",
+        "receipts",
+        "typing",
+        "to_device",
+    ]:
+        return False
+    else:
+        return True
+
+
+def increment_counter(worker_type_counter: Dict[str, int], worker_type: str) -> int:
+    """Given a dict of worker_type:int, increment int
+
+    Args:
+        worker_type_counter: dict to increment
+        worker_type: str of worker_type
+    Returns: int of new value of counter
+    """
+    new_worker_count = worker_type_counter.setdefault(worker_type, 0) + 1
+    worker_type_counter[worker_type] = new_worker_count
+    return new_worker_count
+
+
+def is_name_allowed_for_worker(
+    worker_type_counter: Dict[str, str], worker_base_name: str, worker_type: str
+) -> bool:
+    """Given a dict of worker_base_name:worker_type, check if this worker_base_name has
+        been seen before.
+
+    Args:
+        worker_type_counter: dict of worker_base_name:worker_type.
+        worker_base_name: str of the base worker name, no appended number.
+        worker_type: str of the worker_type, including combo workers.
+    Returns: True if allowed, False if not
+    """
+    counter_to_check = worker_type_counter.get(worker_base_name)
+    if counter_to_check is not None:
+        if counter_to_check == worker_type:
+            # Key exists, and they match so it should be ok.
+            return True
```

Comment:
I'm not sure what this is doing for us or why it should be allowed

Reply:
This checks if the name has been registered to a type of worker. For example,
```
bob=pusher:2
```
If bob is not found then the name is ok.
If bob is found and it's a pusher then the name is ok.
If bob is found but is not a pusher then it's not ok.

I probably could have written more better 😄 I'll see what I can do to make it clearer. Some of the variable names will be changed.

Reply:
I know more now than I did when I wrote this. This can be a simple compound `if` statement, therefore I will remove the function altogether and just hard code it in below.